AsyncResultSetImplTest.testCallbackIsNotCalledWhilePausedAndCanceled is flaky

```
Error:  Tests run: 15, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.211 s <<< FAILURE! -- in com.google.cloud.spanner.AsyncResultSetImplTest
Error:  com.google.cloud.spanner.AsyncResultSetImplTest.testCallbackIsNotCalledWhilePausedAndCanceled -- Time elapsed: 0.005 s <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<2>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.google.cloud.spanner.AsyncResultSetImplTest.testCallbackIsNotCalledWhilePausedAndCanceled(AsyncResultSetImplTest.java:403)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/AsyncResultSetImplTest.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/AsyncResultSetImplTest.java
@@ -104,7 +104,7 @@ public class AsyncResultSetImplTest {
         new AsyncResultSetImpl(
             mockedProvider, mock(ResultSet.class), AsyncResultSetImpl.DEFAULT_BUFFER_SIZE)) {
       rs.setCallback(mock(Executor.class), mock(ReadyCallback.class));
-      IllegalStateException e = assertThrows(IllegalStateException.class, () -> rs.tryNext());
+      IllegalStateException e = assertThrows(IllegalStateException.class, rs::tryNext);
       assertThat(e.getMessage()).contains("tryNext may only be called from a DataReady callback.");
     }
   }
@@ -152,7 +152,7 @@ public class AsyncResultSetImplTest {
   }
 
   @Test
-  public void toListAsyncPropagatesError() throws InterruptedException {
+  public void toListAsyncPropagatesError() {
     ExecutorService executor = Executors.newFixedThreadPool(1);
     ResultSet delegate = mock(ResultSet.class);
     when(delegate.next())
@@ -326,10 +326,7 @@ public class AsyncResultSetImplTest {
               @Override
               public Boolean answer(InvocationOnMock invocation) throws Throwable {
                 row++;
-                if (row > simulatedRows) {
-                  return false;
-                }
-                return true;
+                return row <= simulatedRows;
               }
             });
     when(delegate.getCurrentRowAsStruct()).thenReturn(mock(Struct.class));
@@ -345,17 +342,17 @@ public class AsyncResultSetImplTest {
                 assertFalse(paused.get());
                 callbackCounter.incrementAndGet();
                 try {
-                  while (true) {
-                    switch (resultSet.tryNext()) {
-                      case OK:
-                        paused.set(true);
-                        queue.put(new Object());
-                        return CallbackResponse.PAUSE;
-                      case DONE:
-                        return CallbackResponse.DONE;
-                      case NOT_READY:
-                        return CallbackResponse.CONTINUE;
-                    }
+                  switch (resultSet.tryNext()) {
+                    case OK:
+                      paused.set(true);
+                      queue.put(new Object());
+                      return CallbackResponse.PAUSE;
+                    case DONE:
+                      return CallbackResponse.DONE;
+                    case NOT_READY:
+                      return CallbackResponse.CONTINUE;
+                    default:
+                      throw new IllegalStateException();
                   }
                 } catch (InterruptedException e) {
                   throw SpannerExceptionFactory.propagateInterrupt(e);
@@ -384,9 +381,8 @@ public class AsyncResultSetImplTest {
   }
 
   @Test
-  public void testCallbackIsNotCalledWhilePausedAndCanceled()
-      throws InterruptedException, ExecutionException {
-    Executor executor = Executors.newSingleThreadExecutor();
+  public void testCallbackIsNotCalledWhilePausedAndCanceled() {
+    ExecutorService executor = Executors.newSingleThreadExecutor();
     StreamingResultSet delegate = mock(StreamingResultSet.class);
 
     final AtomicInteger callbackCounter = new AtomicInteger();
@@ -414,6 +410,8 @@ public class AsyncResultSetImplTest {
       SpannerException exception = assertThrows(SpannerException.class, () -> get(callbackResult));
       assertEquals(ErrorCode.CANCELLED, exception.getErrorCode());
       assertEquals(1, callbackCounter.get());
+    } finally {
+      executor.shutdown();
     }
   }
 

